nucleus/async_utils.py

```diff
@@ -0,0 +1,211 @@
+import asyncio
+import time
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, BinaryIO, Callable, Sequence, Tuple
+
+import aiohttp
+import nest_asyncio
+from tqdm import tqdm
+
+from nucleus.constants import DEFAULT_NETWORK_TIMEOUT_SEC
+from nucleus.errors import NucleusAPIError
+from nucleus.retry_strategy import RetryStrategy
+
+from .logger import logger
+
+if TYPE_CHECKING:
+    from . import NucleusClient
+
+
+@dataclass
+class FileFormField:
+    name: str
+    filename: str
+    value: BinaryIO
+    content_type: str
+
+
+FileFormData = Sequence[FileFormField]
+
+
+async def gather_with_concurrency(n, *tasks):
+    semaphore = asyncio.Semaphore(n)
+
+    async def sem_task(task):
+        async with semaphore:
+            return await task
+
+    return await asyncio.gather(*(sem_task(task) for task in tasks))
+
+
+class FormDataContextHandler:
+    """A context handler for file form data that handles closing all files in a request.
+
+    Why do I need to wrap my requests in such a funny way?
+
+    1. Form data must be regenerated on each request to avoid errors
+        see https://github.com/Rapptz/discord.py/issues/6531
+    2. Files must be properly open/closed for each request.
+    3. We need to be able to do 1/2 above multiple times so that we can implement retries
+        properly.
+
+    Write a function that returns a tuple of form data and file pointers, then pass it to the
+    constructor of this class, and this class will handle the rest for you.
+    """
+
+    def __init__(
+        self,
+        form_data_and_file_pointers_fn: Callable[
+            ..., Tuple[FileFormData, Sequence[BinaryIO]]
+        ],
+    ):
+        self._form_data_and_file_pointer_fn = form_data_and_file_pointers_fn
+        self._file_pointers = None
+
+    def __enter__(self):
+        (
+            file_form_data,
+            self._file_pointers,
+        ) = self._form_data_and_file_pointer_fn()
+        form = aiohttp.FormData()
+        for field in file_form_data:
+            form.add_field(
+                name=field.name,
+                filename=field.filename,
+                value=field.value,
+                content_type=field.content_type,
+            )
+        return form
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        for file_pointer in self._file_pointers:
+            file_pointer.close()
+
+
+def get_event_loop():
+    try:
+        loop = asyncio.get_event_loop()
+    except RuntimeError:  # no event loop running:
+        loop = asyncio.new_event_loop()
+    else:
+        nest_asyncio.apply(loop)
+    return loop
+
+
+def make_many_form_data_requests_concurrently(
+    client: "NucleusClient",
+    requests: Sequence[FormDataContextHandler],
+    route: str,
+    progressbar: tqdm,
+    concurrency: int = 30,
+):
```

Comment:
This sounds like a method that would be nice to be able to tune as a user. Depending on what you are uploading this might out-of-mem you or you could increase concurreny in high-mem situtations with a large bandwidth.

Are we exposing the concurrency params to the user at this point?

Reply:
Yes, see the user-facing documentation for how to use/tune this.

Reply:
dataset.append and dataset.annotate will support tuning this param after this PR.